#!/bin/sh
##ID=trim align filter cluster reformat variant_call annotate, Description="Script required to run the program"
##ID=source, Description="Directory where raw data(fq/fq.gz) is located"
##ID=sample_names, Description="Samples list in source"
##ID=tail_name, Description="Name of the tail of each sampleâ€œ
##ID=out_dir, Description="Directory after trim"
##ID=bwaPath samtoolsPath smcounterPath bedtoolsPath, Description="Paths of the tools used"
##ID=ref_fa_file, Description="Target region file(.fasta)"
##ID=index_name, Description="Index file of ref_fa_file"
##ID=primers, Description="Primers target file(.csv)"
##ID=min_consolidate_qual, Description="Minimum quality value in cluster"
##ID=min_consolidate_freq, Description="Minimum frequency in cluster"
##ID=bedTarget, Description="Bed target file(.bed)"
##ID=mtDepth, Description="Mean MT depth in smCounter"
##ID=rpb, Description="Mean read pairs per MT in smCounter"
##ID=nCPU, Description="Number of CPUs to use in parallel in smCounter"
##ID=minBQ, Description="Minimum base quality allowed for analysis in smCounter"
##ID=minMQ, Description="Minimum mapping quality allowed for analysis in smCounter"
##ID=hplen, Description="Minimum length of homopolymers in smCounter"
##ID=mismatchThr, Description="Average number of mismatches per 100 bases allowed in smCounter"
##ID=mtDrop, Description="Drop MTs with lower than or equal to X reads. in smCounter"
##ID=threshold, Description="Minimum prediction index for a variant to be called in smCounter.Must be non-negative. Typically ranges from 10 to 60.If set to 0 (default), smCounter will choose the appropriate cutoff based on the mean MT depth."
##ID=refGenome, Description="RefGenome,select hg19"
##ID=bedTandemRepeats, Description="Bed for UCSC tandem repeats"
##ID=bedRepeatMaskerSubset, Description="Bed for RepeatMasker simple repeats, low complexity,microsatellite regions"
##ID=cosmic clinvar g1000, Description="Databases used to annotate"
##ID=ref_ens, Description="reference Ensembl list(.csv),e.g.'genename,ENSG_ID,ENST_ID'"

trim="/home/administrator/pipeline/NGS_BRCA_Pipelines-master/trim_reads_qiagen_20180305.py"
align="/home/administrator/pipeline/NGS_BRCA_Pipelines-master/align_reads_qiagen_20180220.py"
filter="/home/administrator/pipeline/NGS_BRCA_Pipelines-master/post_alignment_filter_qiagen_20180220.py"
cluster="/home/administrator/pipeline/NGS_BRCA_Pipelines-master/cluster_barcode_qiagen_20180320.py"
reformat="/home/administrator/pipeline/NGS_BRCA_Pipelines-master/reformat_samfile_qiagen_20180320.py"
variant_call="/home/administrator/pipeline/NGS_BRCA_Pipelines-master/variant_call_qiagen_20180315.py"
annotate="/home/administrator/pipeline/NGS_BRCA_Pipelines-master/annotation_qiagen_20180322.py"

source="/home/administrator/sample/"
sample_names="S0721_05B_CHG029767-YN-171205-N701-TAAGGCGA S0721_05B_CHG029767-YN-171205-N703-AGGCAGAA S0721_05B_CHG029767-YN-171205-N707-CTCTCTAC S0721_05B_CHG029767-YN-171205-N711-AAGAGGCA S0721_05B_CHG029767-YN-171205-N715-ATCTCAGG"
tail_name="L003"
out_dir="undetermined/"
bwaPath="bwa"
index_name="/home/administrator/source/refseq"
ref_fa_file="/home/administrator/source/target_breast.refSeq.fa"
samtoolsPath="samtools"
primers="/home/administrator/source/DHS-001Z_primers_target.csv"
min_consolidate_qual=10
min_consolidate_freq=0.8
smcounterPath="/home/administrator/smCounter/smCounter-master/smCounter.py"
bedTarget="/home/administrator/source/DHS-001Z.covered-150bp.bed"
mtDepth="300"
rpb="8.6"
nCPU="16"
minBQ="20"
minMQ="30"
hpLen="8"
mismatchThr="6"
mtDrop="0"
threshold="15"
refGenome="/home/administrator/source/hg19.fa"
bedTandemRepeats="/home/administrator/smCounter/smCounter-master/simpleRepeat.bed"
bedRepeatMaskerSubset="/home/administrator/smCounter/smCounter-master/SR_LC_SL.nochr.bed"
bedtoolsPath="/usr/bin/"
cosmic="/home/administrator/database/COSMIC_variant.csv"
clinvar="/home/administrator/database/breast_cancer_variant_clinvar.csv"
g1000="/home/administrator/database/breast_cancer_variant_1000genomes.csv"
ref_ens="/home/administrator/source/geneid_cancer_qiagen.csv"

for sample_name in $sample_names
do
  command_trim="$trim $source $sample_name $tail_name $out_dir"
  command_align="$align $source $sample_name $bwaPath $index_name $ref_fa_file"
  command_post_alignment="$filter $source $sample_name $samtoolsPath $primers"
  command_cluster="$cluster $source $sample_name $min_consolidate_qual $min_consolidate_freq"
  command_reformat_sam="$reformat $source $sample_name"
  command_variant_call="$variant_call $source $sample_name $samtoolsPath $smcounterPath $bedTarget $mtDepth $rpb $nCPU $minBQ $minMQ $hpLen $mismatchThr $mtDrop $threshold $refGenome $bedTandemRepeats $bedRepeatMaskerSubset $bedtoolsPath"
  command_annotate="$annotate $source $sample_name $cosmic $cilnvar $g1000 $ref_ens"
  
  echo "0 Trim"
  python3 $command_trim
  echo "1 Alignment"
  python3 $command_align
  echo "2 Post alignment analysis"
  python3 $command_post_alignment
  echo "3 Clustering based on UMIs"
  python3 $command_cluster
  echo "4 Reformat SAM files"
  python3 $command_reformat_sam
  echo "5 Variant call"
  python3 $command_variant_call
  echo "6 Annotate"
  python3 $command_annotate
  echo "----------------------------------------------------------"
  echo "sample $sample_name is OK."
  echo "----------------------------------------------------------"
done
